HW26/HW26.cpp

```diff
@@ -0,0 +1,163 @@
+#include <iostream>
+#include <cmath>
+
+enum class VectorRelativeState {
+    Identical,
+    CoDirected,
+    OppositeDirected,
+    AcuteAngle,
+    ObtuseAngle,
+    RightAngle
+};
+
+class Vector2d {
+private:
+    float x, y;
+
+public:
+    Vector2d() : x(0.0f), y(0.0f) {}
+    Vector2d(float x, float y) : x(x), y(y) {}
+
+    Vector2d& operator=(const Vector2d& other) {
+        if (this != &other) {
+            x = other.x;
+            y = other.y;
+        }
+        return *this;
+    }
+
+    Vector2d operator+(const Vector2d& secondVector) const {
+        return Vector2d(x + secondVector.x, y + secondVector.y);
+    }
+
+    Vector2d operator-(const Vector2d& secondVector) const {
+        return Vector2d(x - secondVector.x, y - secondVector.y);
+    }
+
+    friend Vector2d operator+(const Vector2d& leftVector, const Vector2d& rightVector);
+
+    float operator()() const {
+        return std::sqrt(x * x + y * y);
+    }
+
+    float getX() const {
+        return x;
+    }
+
+    float getY() const {
+        return y;
+    }
+
+    friend std::ostream& operator<<(std::ostream& os, const Vector2d& vector) {
+        os << "{" << vector.x << "; " << vector.y << "}";
+        return os;
+    }
+
+    friend std::istream& operator>>(std::istream& is, Vector2d& vector) {
+        is >> vector.x >> vector.y;
+        return is;
+    }
+
+    static int activeInstances;
+
+    static int getActiveInstances() {
+        return activeInstances;
+    }
+
+    ~Vector2d() {
+        activeInstances--;
+    }
+
+    Vector2d(float x0, float y0, float x1, float y1)
+            : x(x1 - x0), y(y1 - y0) {}
+
+    float dotProduct(const Vector2d& other) const {
+        return x * other.x + y * other.y;
+    }
+
+    Vector2d negate() const {
+        return Vector2d(-x, -y);
+    }
+
+    VectorRelativeState getRelativeState(const Vector2d& other) const {
+        float dotProd = dotProduct(other);
+        float mag1 = (*this)();
+        float mag2 = other();
+
+        if (dotProd == 0.0f) {
```

Comment:
comparing float is not that simple in general. You have to compare them up to some presicion. https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon